### PR TITLE
[Contract API] Add contract APIs and add contract consumer

### PR DESCRIPTION
**Pull request by PhyrexTsai**
#### Resolves
- Add contract APIs
```
POST /ethereum/v1/contract
GET /ethereum/v1/contract?contractId=
```
- Add contract consumer
```
/schedulers/aws/sqsContractConsumer.js
```

#### Test cases
- /tests/helpers/aws/sqsHelperTest.js
    - Test send message to AWS SQS, and use FIFO queue
- /tests/others/solcTest.js
    - Test save contract data into Postgers `Contract` table
- /tests/others/deployContractTest.js
    - Test deploy contract to ethereum

### CODE REVIEW

File: models/eventType.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+EventType.prototype.readAll = function() {
+    return pool.query('SELECT * FROM eventtype');
+};
+
+EventType.prototype.read = function(id) {
+    return pool.query('SELECT * FROM eventtype WHERE id = $1', [id]);
+};
+
+EventType.prototype.create = function(id) {
+    return pool.query('SELECT * FROM eventtype WHERE id = $1', [id]);
```
Comment (YanlongLai) [inline models/eventType.js:14]: `create` function is same with `read`?

File: routes/ethereum.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                abi: abi,
+                // have to check [solc]
+                gasEstimates: web3.eth.estimateGas({data: result.contracts[contractName].bytecode})
+            };
+
+            contract.create(contractEntity).then(function (contractId) {
+                id.push(contractId);
+
+                JSON.parse(abi).forEach(function(data){
+                    if (data.type == 'event') {
```
Comment (YanlongLai) [inline routes/ethereum.js]: `===` better than `==`, you can try it in JS
`!==` better than `!=`

File: controllers/ethereumController.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+EthereumController.prototype.getCoinbase = function() {
+	return web3.eth.coinbase;
+}
+
+EthereumController.prototype.getBalance = function(coinbase) {
+	return web3.fromWei(web3.eth.getBalance(coinbase), "ether").toString(10) + ' ether';
+}
+
+exports = module.exports = new EthereumController();
```
Comment (YanlongLai) [inline controllers/ethereumController.js]: github prefer add a blank line in the last line

File: models/contract.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	    return pool.query("INSERT INTO contract (id, name, sourceCode, byteCode, language, compilerVersion, abi, createTimestamp, gasEstimates, status) VALUES " +
+	        "($1, $2, $3, $4, $5, $6, $7, now(), $8, $9)",
+	        [id, entity.name, entity.sourceCode, entity.byteCode, entity.language, entity.compilerVersion, entity.abi, entity.gasEstimates, Contract.prototype.UNCONFIRMED])
+	    .then(function(){
+	        return id;
+	    }).catch(function (err) {
+            console.log(err.message, err.stack);
+        });
+	}).catch(function (err) {
+		console.log(err.message, err.stack);
```
Comment (YanlongLai) [inline models/contract.js]: indent error?

File: models/contract.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+Contract.prototype.updateAddress = function(entity) {
+	return pool.query("UPDATE contract SET address = $1, status = $2, createTimestamp = now() WHERE id = $3", 
+		[entity.address, Contract.prototype.CONFIRMED, entity.id]);
+}
+
+Contract.prototype.delete = function() {
+
+};
+
+exports = module.exports = new Contract();
```
Comment (YanlongLai) [inline models/contract.js]: github prefer add a blank line in the last line

File: models/contractEvent.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+ContractEvent.prototype.update = function() {
+
+};
+
+ContractEvent.prototype.delete = function() {
+
+};
+
+exports = module.exports = new ContractEvent();
```
Comment (YanlongLai) [inline models/contractEvent.js]: github prefer add a blank line in the last line

File: models/contractFunction.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+ContractFunction.prototype.update = function() {
+
+};
+
+ContractFunction.prototype.delete = function() {
+
+};
+
+exports = module.exports = new ContractFunction();
```
Comment (YanlongLai) [inline models/contractFunction.js]: github prefer add a blank line in the last line

File: models/eventData.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+EventData.prototype.update = function() {
+
+};
+
+EventData.prototype.delete = function() {
+
+};
+
+exports = module.exports = new EventData();
```
Comment (YanlongLai) [inline models/eventData.js]: github prefer add a blank line in the last line

File: models/eventFilter.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+EventFilter.prototype.readAll = function() {
+    return pool.query('SELECT * FROM eventfilter');
+};
+
+EventFilter.prototype.read = function(id) {
+    return pool.query('SELECT * FROM eventfilter WHERE id = $1', [id]);
+};
+
+EventFilter.prototype.create = function(id) {
+    return pool.query('SELECT * FROM eventfilter WHERE id = $1', [id]);
```
Comment (YanlongLai) [inline models/eventFilter.js:14]: `create` function?

File: public/stylesheets/style.css [hunk 10]
```diff
@@ -0,0 +1,8 @@
+body {
+  padding: 50px;
+  font: 14px "Lucida Grande", Helvetica, Arial, sans-serif;
```
Comment (YanlongLai) [inline public/stylesheets/style.css:3]: special font-family of Lucida Grande?

File: routes/ethereum.js [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+/**
+ * GET block transaction count
+ */
+router.get('/v1/blockTransactionCount', function (req, res, next) {
+    var blockNumberOrString = req.query.block;
+    if (Number.isInteger(blockNumberOrString) || 
+        blockNumberOrString == 'pending' || 
+        blockNumberOrString == 'earliest' || 
+        blockNumberOrString == 'latest') {
```
Comment (YanlongLai) [inline routes/ethereum.js]: I suppose extract the constant like this
```javascript
var PENDING = 'pending';
var EARLIEST = 'earliest';
var LATEST = 'lastest';
if (Number.isInteger(blockNumberOrString) || 
        blockNumberOrString === 'PENDING' || 
        blockNumberOrString === 'EARLIEST' || 
        blockNumberOrString === 'LATEST') {
......
}

```

File: routes/ethereum.js [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        res.json({'data': {'blockTransactionCount': web3.eth.getBlockTransactionCount(req.query.block)}});
+    }
+    res.json({'error': {'code': 104, 'message': 'block number or block status is needed.'}})
+});
+
+module.exports = router;
```
Comment (YanlongLai) [inline routes/ethereum.js]: last line

File: schedulers/aws/sqsContractConsumer.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    region: process.env.AWS_REGION
+});
+ 
+var consumer = Consumer.create({
+  queueUrl: process.env.AWS_CONTRACT_QUEUE_RUL,
+  handleMessage: function (message, done) {
+    var data = JSON.parse(message.Body);
+    done();
+    contract.read(data.contractId).then(function(result){
+      if (result.rowCount > 0) {
```
Comment (YanlongLai) [inline schedulers/aws/sqsContractConsumer.js:22]: is there any condition of `rowCount <==0`?

File: tests/others/deployContractTest.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
+var Web3 = require('web3');
+var web3 = new Web3();
+var eth = web3.eth;
+
+web3.setProvider(new web3.providers.HttpProvider('http://localhost:8545'));
+
+var proofOfTransactionAbi = [{"constant":false,"inputs":[{"name":"txHash","type":"string"}],"name":"getDataHash","outputs":[{"name":"","type":"bytes32"}],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"dataHash","outputs":[{"name":"","type":"bytes32"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"txHash","type":"string"},{"name":"data","type":"string"}],"name":"setData","outputs":[],"payable":false,"type":"function"},{"anonymous":false,"inputs":[{"indexed":false,"name":"from","type":"address"},{"indexed":false,"name":"txHash","type":"string"},{"indexed":false,"name":"dataHash","type":"bytes32"},{"indexed":false,"name":"time","type":"uint256"}],"name":"setDataHashEvent","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"name":"from","type":"address"},{"indexed":false,"name":"txHash","type":"string"},{"indexed":false,"name":"dataHash","type":"bytes32"},{"indexed":false,"name":"time","type":"uint256"}],"name":"getDataHashEvent","type":"event"}];
```
Comment (YanlongLai) [inline tests/others/deployContractTest.js:7]: how about this?
```javascript
var proofOfTransactionAbi = [
{
  "constant": false,
  "inputs": [{
     "name": "txHash",
     "type": "string"}],
  "name": "getDataHash",
  "outputs": [{"name":"","type":"bytes32"}],
  "payable": false,
  "type": "function"
},
{
  "constant": true,
  "inputs":  [],
  "name": "dataHash",
  "outputs": [{
       "name": "",
       "type": "bytes32"
     }],
  "payable": false,
  "type": "function"
},
......
]
```

File: public/stylesheets/style.css (hunk 10 above)
Comment (PhyrexTsai) [reply public/stylesheets/style.css:3]: This is express template, I will change this after finish whole project

File: schedulers/aws/sqsContractConsumer.js (hunk 13 above)
Comment (PhyrexTsai) [reply schedulers/aws/sqsContractConsumer.js:22]: If `rowCount <== 0` it need throw exception

**Pull request merged**